src/dataflow/operators/notificator.rs

```diff
@@ -0,0 +1,136 @@
+    pending: Vec<Capability<T>>,
+    frontier: Vec<MutableAntichain<T>>,
+    available: VecDeque<(Capability<T>, i64)>,
+}
+
+impl<T: Timestamp> Notificator<T> {
+    pub fn new() -> Notificator<T> {
+        Notificator {
+            pending: Vec::new(),
+            frontier: Vec::new(),
+            available: VecDeque::new(),
+        }
+    }
+
+    /// Updates the `Notificator`'s frontiers from a `CountMap` per input.
+    pub fn update_frontier_from_cm(&mut self, count_map: &mut [CountMap<T>]) {
+        while self.frontier.len() < count_map.len() {
+            self.frontier.push(MutableAntichain::new());
+        }
+
+        for (index, counts) in count_map.iter_mut().enumerate() {
+            while let Some((time, delta)) = counts.pop() {
+                self.frontier[index].update(&time, delta);
+            }
+        }
+    }
+
+    /// Reveals the elements in the frontier of the indicated input.
+    pub fn frontier(&self, input: usize) -> &[T] {
+        self.frontier[input].elements()
+    }
+
+    /// Requests a notification at the time associated with capability `cap`. Takes ownership of
+    /// the capability.
+    ///
+    /// In order to request a notification at future timestamp, obtain a capability for the new
+    /// timestamp first, as show in the example.
+    ///
+    /// #Examples
+    /// ```
+    /// use timely::dataflow::operators::{ToStream, Unary};
+    /// use timely::dataflow::channels::pact::Pipeline;
+    ///
+    /// timely::example(|scope| {
+    ///     (0..10).to_stream(scope)
+    ///            .unary_notify(Pipeline, "example", Vec::new(), |input, output, notificator| {
+    ///                while let Some((cap, data)) = input.next() {
+    ///                    output.session(&cap).give_content(data);
+    ///                    let mut time = cap.time();
+    ///                    time.inner += 1;
+    ///                    notificator.notify_at(cap.delayed(&time));
+    ///                }
+    ///                while let Some((cap, count)) = notificator.next() {
+    ///                    println!("done with time: {:?}", cap.time());
+    ///                }
+    ///            });
+    /// });
+    /// ```
+    #[inline]
+    pub fn notify_at(&mut self, cap: Capability<T>) {
+        self.pending.push(cap);
+    }
+
+    /// Repeatedly calls `logic` till exhaustion of the available notifications.
+    ///
+    /// `logic` receives a capability for `t`, the timestamp being notified and a `count`
+    /// representing how many capabilities were requested for that specific timestamp.
+    #[inline]
+    pub fn for_each<F: FnMut(Capability<T>, i64)>(&mut self, mut logic: F) {
+        while let Some((cap, count)) = self.next() {
+            ::logging::log(&::logging::GUARDED_PROGRESS, true);
+            logic(cap, count);
+            ::logging::log(&::logging::GUARDED_PROGRESS, false);
+        }
+    }
+
+    #[inline]
+    fn scan_for_available_notifications(&mut self) {
```

Comment:
> But, I think if you scan pending for the index of an available capability and update the index if you find a capability less than your current one, that should guarantee that they are delivered not-out-of-order. Right, maybe?

Yup, sorry, I wasn't sure whether you meant that your code was preserving the guarantee (by scanning the entire list) or was just grabbing the first available notification.
I've came up with a (maybe) similar approach, https://github.com/utaal/timely-dataflow/blob/ced70769e866d5d756e7da0657757aa24f5c03e0/src/dataflow/operators/notificator.rs
Can we go with something like that?

> Edit: Also, I think `queue` works without order constraints on notifications. It keeps each queue differentiated by time. What am I missing?

I may have misunderstood what `queue` is supposed to do, but if it can be used to ensure that data is sent in-order to its downstream operators then I think out-of-order delivery of notifications would break it.